### PR TITLE
Exclude fire from exponential decay

### LINKED ISSUE #45318: Fires immediately go out after being lit
# Describe the bug

Fires lit in fireplaces and furnaces go out in a single turn with plenty of perfectly good fuel in Build version: 0.E-7472-g7f7deed
They do not do this in 
Build version: 0.E-7427-g895bbd8 that I was using yesterday.
The log says the fire was successfully lit.

# Steps To Reproduce

go to a lovely ol' furnace in a basement, or a fireplace
use that good ol' advanced inventory to throw some planks and even rags in for good measure
have a lighter or matches
light fire, get told you were successful
fire goes out after a single turn.

# Expected behavior

fire burns as per normal

# Screenshots

# Versions and configuration

Platform     : Windows
Graphics     : Tiles
Build number : #11153
Build version: 0.E-7472-g7f7deed
Major version: 0.E

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
SUMMARY: None

#### Purpose of change
Fixes #45318 
Prevents instantinguishing.

#### Describe the solution
Reverts decay logic for fire to old system, which guarantees that if age is less than half of half_life, decay will not happen.

#### Describe alternatives you've considered
Fire logic in general is incredibly convoluted for no good reason and needs a complete overhaul, but not today.

#### Testing
Lit a pile of zombie junk on fire, and the fire stuck around instead of immediately extinguishing.

#### Additional context
I'm working on a unit test for field lifetimes, but I don't want to block this quick fix behind that.

### CODE REVIEW

File: src/field.cpp [hunk 1]
```diff
@@ -152,6 +153,13 @@ void field_entry::do_decay()
     // Bypass set_field_age() so we don't reset decay_time;
     age += 1_turns;
     if( type.obj().half_life > 0_turns && get_field_age() > 0_turns ) {
+        // Legacy handling for fire because it's weird and complicated.
+        if( get_field_type() == STATIC( field_type_str_id( "fd_fire" ) ) ) {
+            if( to_turns<int>( fdata.half_life ) < dice( 2, to_turns<int>( cur.get_field_age() ) ) ) {
```
Comment (anothersimulacrum) [inline src/field.cpp:158]:
```suggestion
            if( to_turns<int>( type->half_life ) < dice( 2, to_turns<int>( age ) ) ) {
```

File: src/field.cpp [hunk 2]
```diff
@@ -152,6 +153,13 @@ void field_entry::do_decay()
     // Bypass set_field_age() so we don't reset decay_time;
     age += 1_turns;
     if( type.obj().half_life > 0_turns && get_field_age() > 0_turns ) {
+        // Legacy handling for fire because it's weird and complicated.
+        if( get_field_type() == STATIC( field_type_str_id( "fd_fire" ) ) ) {
```
Comment (anothersimulacrum) [inline src/field.cpp:157]: This isn't necessary or anything, but there's no strict need to use the function.
```suggestion
        if( type == STATIC( field_type_str_id( "fd_fire" ) ) ) {
```

**Comment by Tamiore** [comment]
Yesterday I was playing on latest experimental and the resulting overall firestarting meta felt rather realistic and very much on point. 

That is, I was only able to start a fire after a dozen or so attempts, but once it "got going", it was easy to maintain. 

Previous meta where you could set anything flammable on fire within a few seconds with just a single match. Yea, THAT feels off.

**Comment by anothersimulacrum** [comment]
Nothing against doing that, but if that is done, it needs to be explicit in the firestarting code, instead of a micro-managey bug where you need to start several fires and wait to see if they catch.

**Comment by Tamiore** [comment]
I agree that it would be best if everything was working as intended, not just by happy accident. 

However, I still like the actual gameplay of this "bugged" meta much better than the previous one. ¯\_(ツ)_/¯

**Comment by wapcaplet** [comment]
Playtesting this, I was able to get fires going in a brazier (or without), using piles of logs, stout branches, planks, or splintered wood - but *only* if I included 10 or more tinder before lighting them. Without tinder, I found I could not get any fires to continue more than 1 or 2 turns, including: rags, sheets, filthy clothes and dead zombies. I also could not light any shrubs or trees (living or dead) on fire for more than 1 turn.

**Comment by kevingranade** [comment]
Reproduced wapletcaplet's issue in a test by lighting a shrub, which burns for as little as 4 turns before going out.
That should make reproduction easier at least.

**Comment by morthein** [comment]
I don't know if this helps or not, but I generally burn things like plastic bags, empty boxes, and other trash along with 2x4's and logs in my wood stove.  I noticed that every time I light it, it will only burn up one item category before going out.
In a stove with 5 different fire sources, I have to light it 5 times. (and sometimes when it triggers the fuel made from wood, it changes it to the "burnt" version and goes out.)

Also, sometimes when you light a fire, you get a notice that shows you "you successfully lit a fire" but no flame actually appears, although it does remove a use from your chosen fire source.

**Comment by kevingranade** [comment]
Well that was dumb, I added the special handling for fire fields, but forgot to add anything to skip the non-fire logic, so the non-fire updating was still applying to fire fields.
Also added some field lifetime tests, and based on failures there, adjusted decay rate calculation as well.

**Comment by fizzle7** [comment]
I feel like exponential decay will be a problem even after fire is fixed.
I don't think random short fire duration should be deemed "realistic" to simulate when you fail to light a fire in real life. This should be done through a proper mechanism (wind, tinder quality, fuel, etc.).
The bigger problem is, every field under exponential decay (and fire, when it's fixed and standardized under it) could have a short duration or long duration, and I'm not sure that it's not a problem. There is a somewhat high chance that you might get several times the half-life of a field as its lifetime (got as much as 6 times the half-life of fire in my tests), or 0.1 or lower times the half-life (got as low as 5 turns for fire). I think it would be better if the exponential decay had hard limits, or if it was modified somehow, perhaps a normal or a custom distribution based on the exponential.

**Comment by kevingranade** [comment]
Fire is getting an exemption, not just because of the bug, but because exponential decay doesn't make sense for fire at all, so you're overthinking things.
Other fields are perfectly fine with exponential decay.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/fire-fireplaces/25337/7

**Comment by dissociativity** [comment]
am I missing something in the checks? the 'failed' ones seem to imply the process was cancelled rather than any actual test failures

File: tests/field_test.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+#include "map_helpers.h"
+
+#include <algorithm>
+
+static int count_fields( field_type_str_id &field_type )
+{
+    map &m = get_map();
+    int live_fields = 0;
+    for( tripoint cursor : m.points_on_zlevel() ) {
```
Comment (KorGgenT) [inline tests/field_test.cpp:16]:
```suggestion
    for( const tripoint &cursor : m.points_on_zlevel() ) {
```

File: tests/field_test.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// This is a large scale integration test, it gets prohibitively expensive if it uses
+// a long-lived field, and it isn't capable of handling fields that trigger spread/join mechanics
+// or dynamically update their own lifetimes.
+TEST_CASE( "acid_field_expiry_on_map", "[field]" )
+{
+    clear_map();
+    map &m = get_map();
+    field_type_str_id field_type( "fd_acid" );
```
Comment (KorGgenT) [inline tests/field_test.cpp:31]:
```suggestion
    const field_type_str_id field_type( "fd_acid" );
```

File: tests/field_test.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    // place a smoke field
+    for( tripoint cursor : m.points_on_zlevel() ) {
+        m.add_field( cursor, field_type, 1 );
+    }
+    REQUIRE( count_fields( field_type ) == 17424 );
+    time_point before_time = calendar::turn;
```
Comment (KorGgenT) [inline tests/field_test.cpp:37]:
```suggestion
    const time_point before_time = calendar::turn;
```

File: tests/field_test.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    test_field_expiry( "fd_acid" );
+    test_field_expiry( "fd_smoke" );
+    test_field_expiry( "fd_blood" );
+    test_field_expiry( "fd_sludge" );
+    test_field_expiry( "fd_extinguisher" );
+    test_field_expiry( "fd_electricity" );
+}
+
+static void fire_duration( const std::string &terrain_type, time_duration minimum,
+                           time_duration maximum )
```
Comment (KorGgenT) [inline tests/field_test.cpp:83]:
```suggestion
static void fire_duration( const std::string &terrain_type, const time_duration minimum,
                           const time_duration maximum )
```

File: tests/field_test.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+{
+    clear_map();
+    tripoint fire_loc{ 33, 33, 0 };
```
Comment (KorGgenT) [inline tests/field_test.cpp:86]:
```suggestion
    const tripoint fire_loc{ 33, 33, 0 };
```

File: tests/field_test.cpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    m.ter_set( fire_loc, ter_id( terrain_type ) );
+    m.add_field( fire_loc, fd_fire, 1, 10_minutes );
+    REQUIRE( m.get_field( fire_loc, fd_fire ) );
+    CHECK( m.get_field( fire_loc, fd_fire )->is_field_alive() );
+    time_point before_time = calendar::turn;
+    bool field_alive = true;
+    while( field_alive && calendar::turn - before_time < minimum ) {
+        m.process_fields();
+        calendar::turn += 1_seconds;
+        int effective_age = to_turns<int>( calendar::turn - before_time );
```
Comment (KorGgenT) [inline tests/field_test.cpp:98]:
```suggestion
        const int effective_age = to_turns<int>( calendar::turn - before_time );
```

File: tests/field_test.cpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+{
+    clear_map();
+    tripoint fire_loc{ 33, 33, 0 };
+    map &m = get_map();
+    //m.ter_set( fire_loc, ter_id( "t_shrub_raspberry" ) );
+    m.ter_set( fire_loc, ter_id( terrain_type ) );
+    m.add_field( fire_loc, fd_fire, 1, 10_minutes );
+    REQUIRE( m.get_field( fire_loc, fd_fire ) );
+    CHECK( m.get_field( fire_loc, fd_fire )->is_field_alive() );
+    time_point before_time = calendar::turn;
```
Comment (KorGgenT) [inline tests/field_test.cpp:93]:
```suggestion
    const time_point before_time = calendar::turn;
```

File: tests/field_test.cpp [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        int effective_age = to_turns<int>( calendar::turn - before_time );
+        INFO( effective_age << " seconds" );
+        field_entry *this_field = m.get_field( fire_loc, fd_fire );
+        field_alive = this_field && this_field->is_field_alive();
+    }
+    CHECK( field_alive );
+    while( field_alive && calendar::turn - before_time < maximum ) {
+        m.process_fields();
+        calendar::turn += 1_seconds;
+        int effective_age = to_turns<int>( calendar::turn - before_time );
```
Comment (KorGgenT) [inline tests/field_test.cpp:107]:
```suggestion
        const int effective_age = to_turns<int>( calendar::turn - before_time );
```

**Review by KorGgenT — COMMENTED** [review]
just some low hanging `const`s

**Comment by KorGgenT** [comment]
when i light a fire on a plank or a log on the ground, it catches, but after waiting 30 minutes to an hour, it burns out without burning up the log or plank at all.

File: tests/field_test.cpp [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    test_field_expiry( "fd_electricity" );
+}
+
+static void fire_duration( const std::string &terrain_type, const time_duration minimum,
+                           const time_duration maximum )
+{
+    clear_map();
+    const tripoint fire_loc{ 33, 33, 0 };
+    map &m = get_map();
+    //m.ter_set( fire_loc, ter_id( "t_shrub_raspberry" ) );
```
Comment (ZhilkinSerg) [inline tests/field_test.cpp:88]:
```suggestion
    //m.ter_set( fire_loc, ter_id( "t_shrub_raspberry" ) );
```

File: tests/field_test.cpp [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+#include "field.h"
+#include "map.h"
+#include "map_iterator.h"
+#include "type_id.h"
+
+#include "map_helpers.h"
+
+#include <algorithm>
+
+static int count_fields( field_type_str_id &field_type )
```
Comment (ZhilkinSerg) [inline tests/field_test.cpp:12]:
```suggestion
static int count_fields( const field_type_str_id &field_type )
```

File: tests/field_test.cpp [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+// This is a large scale integration test, it gets prohibitively expensive if it uses
+// a long-lived field, and it isn't capable of handling fields that trigger spread/join mechanics
+// or dynamically update their own lifetimes.
+TEST_CASE( "acid_field_expiry_on_map", "[field]" )
+{
+    clear_map();
+    map &m = get_map();
+    const field_type_str_id field_type( "fd_acid" );
+    // place a smoke field
+    for( tripoint cursor : m.points_on_zlevel() ) {
```
Comment (ZhilkinSerg) [inline tests/field_test.cpp:33]:
```suggestion
    for( const tripoint &cursor : m.points_on_zlevel() ) {
```

**Pull request merged**